[orchagent] CoPP neighbor miss trap and enhancements

rminnikanti opened the pull request:
**What I did**

- Added neighbor_miss trap type support
- enum capability query for hostif trap type in copporch.cpp
- Added hw_status field to COPP_TRAP_TABLE table in state_db

**Why I did it**

HLD: [sonic-net/SONiC#1943](https://github.com/sonic-net/SONiC/pull/1943)

**How I verified it**

Verified with test_copp.py DVS test
Verified on Sonic switch with and without neighbor_miss trap support by vendorSAI.

**Details if related**

PR Dependencies:

- Depends on: sonic-net/sonic-swss-common#1010
- Depends on: sonic-net/sonic-sairedis#1586

Signed-off-by: Ravi Minnikanti <rminnikanti@marvell.com>

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

rminnikanti commented:
@prsunny @kcudnik PR check compilation failing because of dependency on sonic-swss-common PR - sonic-net/sonic-swss-common#1010
Can you please suggest if there is a way to run PR checks with dependencies?

kcudnik commented:
there is no way to run code without dependency checked in, technically you could modify azure pipeline to pull swss common at specific commit, compile it and then build against that, but this is not standard approach and it would break everything else

rminnikanti commented:
Thank you for your response. Can the dependency PR be merged to allow the PR checks to run on this PR?
dependency - [sonic-net/sonic-swss-common#1010](https://github.com/sonic-net/sonic-swss-common/pull/1010)

dgsudharsan commented on orchagent/copporch.cpp:
```diff
@@ -89,9 +95,55 @@ static map<string, sai_hostif_trap_type_t> trap_id_map = {
+    {"neighbor_miss", SAI_HOSTIF_TRAP_TYPE_NEIGHBOR_MISS}
 };
 
+const vector<sai_hostif_trap_type_t> default_supported_trap_ids = {
```
As mentioned in the HLD, It is better to install the trap if the query API isn't available. This would ensure we do not have to edit this in the future

rminnikanti replied:
Hi @dgsudharsan,

The default_supported_trap_ids list ensures backward compatibility. It includes all currently known traps used by orchagent, except for "neighbor_miss". This ensures that all existing trap configurations are pushed to the SAI, even if the query API isn't available. It also prevents orchagent from throwing exceptions in cases where the vendor SAI does not support "neighbor_miss" and using a different method to handle neighbor_miss traffic.

In summary, all current traps are pushed to SAI even if query API is not available only neighbor_miss and future traps need query API availability.

dgsudharsan replied:
@prsunny Are you fine with the approach suggested by @rminnikanti ? My concern is if some vendor doesn't support the query but lets assume they support newer traps, in order for it to work, we should either get new vendor SAI with query support, or edit orchagent default_supported_trap_ids. If this is not a problem, I am fine with the approach listed

rminnikanti replied:
@msherif1234, were you suggesting moving it to copporch.h or creating a separate header file specifically for this vector? Including it in copporch.h might unnecessarily expose it to all files that include that header. For reference, aclorch has a similar default list defined in its .cpp file. Let me know your thoughts.

msherif1234 replied:
I was thinking of separate file since this table will be more of static table will never change and new traps will need to go to the main table , this is more of nit not sure what others think about this

rminnikanti replied:
@msherif1234, I agree with the reasoning behind your suggestion. However, to maintain consistency with the existing code style, I've added clear comments to convey the intent.

mssonicbld commented:
/azp run

msherif1234 commented:
do you need to also update mock test json file https://github.com/sonic-net/sonic-swss/pull/3646/files#diff-7a754ce5badfcc6c7f079baabf320c731eab42f9f98194df7ec53eccb0e6064bR106 ?

msherif1234 commented on orchagent/debugcounterorch.cpp:
```diff
@@ -227,7 +227,7 @@ void DebugCounterOrch::doTask(Consumer& consumer)
 // publishDropCounterCapabilities queries the SAI for available drop counter
 // capabilities on this device and publishes the information to the
-// DROP_COUNTER_CAPABILITIES table in STATE_DB.
+// DEBUG_COUNTER_CAPABILITIES table in STATE_DB.
```
nit: why this changed we still dealing with dropCapabilities here not debug ?

mssonicbld commented:
/azp run

rminnikanti commented:
> do you need to also update mock test json file https://github.com/sonic-net/sonic-swss/pull/3646/files#diff-7a754ce5badfcc6c7f079baabf320c731eab42f9f98194df7ec53eccb0e6064bR106 ?

mock_tests/copp_cfg.json is being used in coppmgrd UT.
I still updated it as suggested for consistency.

rminnikanti commented:
> I created PR under sonic-mgmt to add new CoPP configs for this new trap type PTAL [sonic-net/sonic-mgmt#18480](https://github.com/sonic-net/sonic-mgmt/pull/18480)

@msherif1234 I have sonic-mgmt implemented for this HLD which already covers validating neighbor_miss traffic with default 200PPS if the neighbor_miss trap is supported [sonic-net/sonic-mgmt#18326](https://github.com/sonic-net/sonic-mgmt/pull/18326)

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

msherif1234 commented:
/lgtm

prsunny commented on orchagent/copporch.cpp:
```diff
@@ -122,27 +185,118 @@ const string default_trap_group = "default";
+    const auto* meta = sai_metadata_get_attr_metadata(SAI_OBJECT_TYPE_HOSTIF_TRAP, SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE);
+    if (!meta || !meta->isenum) 
+    {
+        SWSS_LOG_THROW("sai_metadata_get_attr_metadata for SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE Failed");
```
Can we change this to WARN and not throw?

rminnikanti replied:
Done

prsunny commented on orchagent/copporch.cpp:
```diff
@@ -122,27 +185,118 @@ const string default_trap_group = "default";
+                                                                     &enum_values_capability);
+    if (status != SAI_STATUS_SUCCESS)
+    {
+        SWSS_LOG_WARN("SAI capability query for SAI_HOSTIF_TRAP_ATTR_TRAP_TYPE is failed,"
```
This can be changed to NOTICE as there could be many cases (like DPU) where this can return NOT_SUPPORTED

prsunny commented on orchagent/copporch.cpp:
```diff
@@ -122,27 +185,118 @@ const string default_trap_group = "default";
+        values_list.assign(default_supported_trap_ids.begin(), default_supported_trap_ids.end());
+        enum_values_capability.list = values_list.data();
+    }
+
```
Can you please add a NOTICE log to print the value of `enum_values_capability.count`

rminnikanti replied:
Done

prsunny commented on orchagent/copporch.cpp:
```diff
@@ -122,27 +185,118 @@ const string default_trap_group = "default";
+        trap_id_list_str += trap_str;
+    }
+
+    SWSS_LOG_INFO("Publishing supported trap IDs to STATE_DB");
```
This can be changed to NOTICE since it only once during init.

prsunny reviewed (commented):
Overall, lgtm. Minor comments on logs.

mssonicbld commented:
/azp run

rminnikanti commented:
> > Addressed review comments. PR checks failed because of dependency on - [sonic-net/sonic-swss-common#1010](https://github.com/sonic-net/sonic-swss-common/pull/1010)
> 
> @rminnikanti PR #1010 is already merged. Why its still a dependency?

@prsunny, unsure why. Below is the error, where it failed to find the macro that is added as part of [sonic-net/sonic-swss-common#1010]
```
../../../orchagent/copporch.cpp: In constructor 'CoppOrch::CoppOrch(swss::DBConnector*, std::string)':
../../../orchagent/copporch.cpp:199:78: error: 'STATE_COPP_TRAP_CAPABILITY_TABLE_NAME' was not declared in this scope; did you mean 'STATE_SWITCH_CAPABILITY_TABLE_NAME'?
  199 |     m_trapCapabilityTable(std::unique_ptr<Table>(new Table(m_state_db.get(), STATE_COPP_TRAP_CAPABILITY_TABLE_NAME))),
      |                                                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                                                              STATE_SWITCH_CAPABILITY_TABLE_NAME
```

mssonicbld commented:
/azp run

prsunny commented:
/azp run Azure.sonic-swss

mssonicbld commented:
/azp run

rminnikanti commented:
@prsunny @prabhataravind 

Can you please help find why PR Build fails even though dependent sonic-swss-common change is merged?
Build unable to find new macro introduced in sonic-swss-common repo, even after the PR is merged.
I am able to build debian locally without any issues. 

```
./../../orchagent/copporch.cpp:199:78: error: 'STATE_COPP_TRAP_CAPABILITY_TABLE_NAME' was not declared in this scope; did you mean 'STATE_SWITCH_CAPABILITY_TABLE_NAME'?
  199 |     m_trapCapabilityTable(std::unique_ptr<Table>(new Table(m_state_db.get(), STATE_COPP_TRAP_CAPABILITY_TABLE_NAME))),
```

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

prsunny commented:
@rminnikanti , can you please check failure?
@prabhataravind , would you please signoff?

mssonicbld commented:
/azp run

rminnikanti commented:
@prsunny, Checks on this PR failing because `libswsscommon-dev_1.0.0_amd64.deb` being downloaded doesn't include my https://github.com/sonic-net/sonic-swss-common/pull/1010 PR.  **sonic-swss-common pipeline is failing since May 28th with libyang3 error**:

This is the libswsscommon-dev debian link being downloaded by this PR build:
https://sonic-build.azurewebsites.net/api/sonic/artifacts?branchName=master&definitionId=9&artifactName=sonic-swss-common-bookworm&target=libswsscommon-dev_1.0.0_amd64.deb

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

mssonicbld commented:
/AzurePipelines run Azure.sonic-swss

prsunny commented:
> @prsunny, PR checks have progressed after the partial success of the sonic-swss-common pipeline. However, one CoPP VS test is currently failing due to the pending merge of PR [#22406](https://github.com/sonic-net/sonic-buildimage/pull/22406) in sonic-buildimage.
> 
> PR check VS test failure log:
> 
> ```shell
> 2025-07-12T05:42:19.7546435Z FAILED test_copp.py::TestCopp::test_defaults - assert False == True
> 2025-07-12T05:42:19.7547029Z =================== 1 failed, 14 passed in 186.25s (0:03:06) ===================
> ```
> 
> Test case verifies whether traps defined in copp_cfg.json are correctly installed. Since PR [#22406](https://github.com/sonic-net/sonic-buildimage/pull/22406) hasn't been merged yet, the neighbor_miss trap is not installed, causing the test to fail as it expects this trap to be present.
> 
> Test case passes when I run locally with sonic-buildimage PR.

hi @rminnikanti , not sure i understand. The tests must be run independently of 22406. We won't be able to merge with PR checker failures

mssonicbld commented:
/azp run

mssonicbld commented:
/azp run

prsunny commented:
@rminnikanti , please don't force-push any commit. Reviewers prefer to understand the incremental changes

prsunny commented:
@dgsudharsan , can you please approve?

Aravind-Subbaroyan commented:
@prsunny - can this be cherry-picked for 202411?

mssonicbld commented:
Cherry-pick PR to 202505: https://github.com/sonic-net/sonic-swss/pull/3811
